File: frontend/src/types/session.ts
// Session 元数据类型 —— 对齐后端 src/api/schemas/session.py
import type { AssistantMessage, Message, ToolCallState } from '@/types/chat'
import { parseUserMessage } from '@/lib/attachments'
import { generateId } from '@/lib/id'

export type Session = {
  id: string
  title: string
  created_at: string
  msg_count: number
}

export type SessionListResponse = {
  sessions: Session[]
}

type BackendToolCall = {
  id?: string
  type?: string
  function?: {
    name?: string
    arguments?: string
  }
}

export type SessionMessagesResponse = {
  messages: Array<{
    id?: number
    user_index?: number
    role: string
    content: string
    tool_calls?: BackendToolCall[]
    tool_call_id?: string
  }>
  has_more: boolean
  oldest_id: number | null
}

/**
 * 把后端 messages 列表（OpenAI 格式）转成前端 Message[]（用户/助手气泡）。
 *
 * 重建策略：
 *   - 一个 user message 后到下一个 user message 之间的所有 assistant + tool messages
 *     合并成单个前端 AssistantMessage（恢复 WorkBlock 视图）
 *   - assistant.tool_calls → ToolCallState[]（args 从 JSON 字符串反解析）
 *   - 紧跟的 role=tool message 的 content → 对应 tool_call 的 preview
 *   - thinking / plan 不重建（流式期间的瞬态信号；plan 当前结构需要 make_plan 工具 args
 *     反推，复杂度高，暂留空。用户要看完整 plan 去"学习计划"view）
 */
export function backendMessagesToFrontend(
  raw: SessionMessagesResponse['messages'],
): Message[] {
  const out: Message[] = []
  let pendingAssistant: AssistantMessage | null = null
  let toolCallIndex = new Map<string, ToolCallState>()

  const finalize = () => {
    if (pendingAssistant) {
      out.push(pendingAssistant)
      pendingAssistant = null
      toolCallIndex = new Map()
    }
  }

  for (const m of raw) {
    if (m.role === 'user') {
      finalize()
      const { text, attachments } = parseUserMessage(m.content)
      out.push({
        id: m.id != null ? String(m.id) : generateId(),
        role: 'user',
        content: text,
        rawContent: m.content,
        attachments,
        userIndex: typeof m.user_index === 'number' ? m.user_index : undefined,
      })
      continue
    }

    if (m.role === 'assistant') {
      if (!pendingAssistant) {
        pendingAssistant = {
          id: generateId(),
          role: 'assistant',
          content: '',
          plan: null,
          timeline: [],
          error: null,
          streaming: false,
        }
      }
      // 累积 tool_calls 到当前轮
      const tcs = m.tool_calls
      if (Array.isArray(tcs)) {
        for (const tc of tcs) {
          const callId = tc.id ?? ''
          const fn = tc.function ?? {}
          const name = fn.name ?? ''
          let args: Record<string, unknown> = {}
          if (fn.arguments) {
            try {
              args = JSON.parse(fn.arguments) as Record<string, unknown>
            } catch {
              args = { _raw: fn.arguments }
            }
          }
          const state: ToolCallState = {
            call_id: callId,
            name,
            args,
            status: 'ok',
          }
          // 把同一个 state 引用放进 timeline，后续 role=tool 行回填 preview 时直接改它
          pendingAssistant.timeline.push({ kind: 'tool', call: state })
          if (callId) toolCallIndex.set(callId, state)
        }
      }
      // 累积 content（同轮内多条 assistant 含 content 时拼接，极少见）
      if (m.content) {
        pendingAssistant.content = pendingAssistant.content
          ? `${pendingAssistant.content}\n\n${m.content}`
          : m.content
      }
      continue
    }

    if (m.role === 'tool') {
      const tid = m.tool_call_id ?? ''
      const state = toolCallIndex.get(tid)
      if (state) {
        state.preview = m.content
      }
      continue
    }
    // role=system 等其它角色不展示
  }
  finalize()
  return out
}

File: src/api/schemas/session.py
"""Session 管理端点请求 / 响应模型"""

from typing import Any

from pydantic import BaseModel, Field


class SessionInfo(BaseModel):
    """list / get 返回的单条 session 元数据"""

    id: str = Field(..., description="会话 ID")
    title: str = Field("", description="会话标题（自动取首条用户消息预览或用户手动改名）")
    created_at: str = Field(..., description="创建时间，ISO 8601 本地时间")
    msg_count: int = Field(0, description="该会话累计消息数（不含 system）")


class SessionListResponse(BaseModel):
    sessions: list[SessionInfo]


class SessionCreateRequest(BaseModel):
    title: str | None = Field(None, description="可选初始标题；不传则为空")


class SessionRenameRequest(BaseModel):
    title: str = Field(..., min_length=1, description="新标题，非空")


class SessionDeleteResponse(BaseModel):
    deleted: bool = Field(..., description="是否实际删除了记录（False = session 不存在）")


class SessionMessagesResponse(BaseModel):
    """拉某 session 的消息历史（支持分页）"""

    messages: list[dict[str, Any]] = Field(
        ...,
        description="OpenAI messages 格式（含 id / user_index / tool_calls 等可选字段）",
    )
    has_more: bool = Field(
        False,
        description="是否还有更早的消息（上滚加载时传 oldest_id 为 before_id）",
    )
    oldest_id: int | None = Field(
        None,
        description="本页最早一条消息的 DB id，作为下一页 before_id 游标",
    )


class SessionTruncateRequest(BaseModel):
    """编辑重发 / 重新生成：从第 user_message_index 条 user 消息起截断"""

    user_message_index: int = Field(
        ..., ge=0, description="第几条 user 消息（0 基），从它起（含）删除后续全部消息"
    )


class SessionTruncateResponse(BaseModel):
    deleted: int = Field(..., description="实际删除的消息行数")
